copy/copy.go

```diff
@@ -151,65 +170,226 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+func (c *copier) copyMultipleImages(policyContext *signature.PolicyContext, options *Options, unparsedToplevel *image.UnparsedImage) (retErr error) {
+	// Parse the list and get a copy of the original value after it's re-encoded.
+	manifestList, manifestType, err := unparsedToplevel.Manifest()
+	if err != nil {
+		return errors.Wrapf(err, "Error reading manifest list")
+	}
+	list, err := manifest.ListFromBlob(manifestList, manifestType)
+	if err != nil {
+		return errors.Wrapf(err, "Error parsing manifest list %q", string(manifestList))
+	}
+	originalList := list.Clone()
+
+	// Read and/or clear the set of signatures for this list.
+	var sigs [][]byte
+	if options.RemoveSignatures {
+		sigs = [][]byte{}
+	} else {
+		c.Printf("Getting image list signatures\n")
+		s, err := c.rawSource.GetSignatures(context.TODO(), nil)
 		if err != nil {
-			return errors.Wrapf(err, "Error choosing an image from manifest list %s", transports.ImageName(srcRef))
+			return errors.Wrap(err, "Error reading signatures")
 		}
-		logrus.Debugf("Source is a manifest list; copying (only) instance %s", instanceDigest)
-		unparsedInstance := image.UnparsedInstance(rawSource, &instanceDigest)
+		sigs = s
+	}
+	if len(sigs) != 0 {
+		c.Printf("Checking if image list destination supports signatures\n")
+		if err := c.dest.SupportsSignatures(); err != nil {
+			return errors.Wrap(err, "Can not copy signatures")
+		}
+	}
 
-		if err := c.copyOneImage(policyContext, options, unparsedInstance); err != nil {
+	// Copy each image, in turn.
+	instanceBlobs := list.Instances()
+	updates := make([]manifest.ManifestListUpdate, len(instanceBlobs))
+	for i, instance := range instanceBlobs {
+		logrus.Debugf("Copying instance %s (%d/%d)", instance.Digest, i+1, len(instanceBlobs))
+		unparsedInstance := image.UnparsedInstance(c.rawSource, &instance.Digest)
+		updatedManifest, updatedManifestType, err := c.copyOneImage(policyContext, options, unparsedInstance, &instance.Digest, &instance.Digest)
+		if err != nil {
+			return err
+		}
+		// Record the result of a possible conversion here.
+		md, err := manifest.Digest(updatedManifest)
+		if err != nil {
 			return err
 		}
+		update := manifest.ManifestListUpdate{
+			Digest:    md,
+			Size:      int64(len(updatedManifest)),
+			MediaType: updatedManifestType,
+		}
+		updates[i] = update
 	}
 
-	if err := c.dest.Commit(); err != nil {
-		return errors.Wrap(err, "Error committing the finished image")
+	// Now apply the updates.
+	if err = list.UpdateInstances(updates); err != nil {
+		return errors.Wrapf(err, "Error updating manifest list")
+	}
+
+	// Check if the updates meaningfully changed the list of images.
+	listIsModified := false
+	if !reflect.DeepEqual(list.Instances(), originalList.Instances()) {
+		listIsModified = true
+	}
+
+	// Determine if we need to convert the manifest to a different format.
+	selectedType, _, err := c.determineManifestListConversion(manifestType, c.dest.SupportedManifestMIMETypes(), "")
```

Comment:
I’m sorry, by “similar to `determineManifestConversion`” I didn’t mean necessary literally equal; including the logic for the `forcedManifestListMIMEType` parameter, which is actually hard-coded to `""` here, and the `otherTypes` return value, which is ignored here, is not useful AFAICT, only confusing dead code.

Reply:
The returned `otherTypes` list is still ignored in the only caller.

Reply:
`determineListConversion` no longer returns that value.